include/TFEL/Material/Eshelby.ixx

```diff
@@ -0,0 +1,283 @@
+ * \copyright Copyright (C) 2006-2018 CEA/DEN, EDF R&D. All rights
+ * reserved.
+ * This project is publicly released under either the GNU GPL Licence
+ * or the CECILL-A licence. A copy of thoses licences are delivered
+ * with the sources of TFEL. CEA or EDF may also distribute this
+ * project under specific licensing conditions.
+ */
+
+#ifndef LIB_TFEL_MATERIAL_ESHELBY_IXX
+#define LIB_TFEL_MATERIAL_ESHELBY_IXX
+
+#include <cmath>
+
+namespace tfel::material
+{
+    /*!
+     * \brief a structure in charge of computing the Eshelby tensor
+     * \tparam real: numeric type
+     */
+	template <unsigned short N,typename real>
+	struct ComputeEshelbyTensor
+	{
+		/* Just Eshelby tensor of a sphere
+		*/
+	      	TFEL_HOST_DEVICE static const tfel::math::st2tost2<3u, real>
+	      	EshelbyTensorSphere(const real& nu)
+	      	{
+	      		const auto zero = real{0};
+		      	const auto J=real{(5.*nu-1.)/15/(1-nu)};
+			const auto I=real{2*(4.-5.*nu)/15/(1-nu)};
+			return {I+J, J, J, 	  zero, zero, zero,
+				J, I+J, J, 	  zero, zero, zero,
+				J, J, I+J, 	  zero, zero, zero,
+				zero, zero, zero, I,  zero, zero,
+				zero, zero, zero, zero, I,  zero,
+				zero, zero, zero, zero, zero, I};
+		};//end EshelbyTensorSphere
+		
+		static constexpr auto eps = std::numeric_limits<real>::epsilon();
+		
+		/*
+		* This function q_ is needed just for the axisymetric ellipsoid
+		*/
+	      	TFEL_HOST_DEVICE static real q_(const real& e) {if (e<1-eps) return e/tfel::math::power<3,2>(1-e*e)*(acos(e)-e*sqrt(1-e*e)); else if (e>1+eps) return e/tfel::math::power<3,2>(e*e-1)*(e*sqrt(e*e-1)-acosh(e)); else return 2./3;};
+		
+		/* EshelbyTensorAxisym is relative to an axisymetric ellipsoid which has two equal semi-axes and another semi-axis which is different
+		* The arguments are nu and the aspect ratio e of the ellipsoid (e>1 : prolate, e<1 : oblate)
+		* The function returns the Eshelby tensor in the basis (e1,e2,e3) where e3 is aligned with the axis of revolution of the ellipsoid
+		* The formulae come from Torquato, Random Heterogeneous Materials, 2002.
+		*/
+		TFEL_HOST_DEVICE static const tfel::math::st2tost2<3u, real>
+		EshelbyTensorAxisym(const real& nu, const real& e)
+		{
+			// if (abs(e-1)<value(eps)){
+			//	the formulae must be adapted in the limit case e->1
+			//			   }
+			const auto zero = real{0};
+			const auto e2=e*e;
+			const auto e21=e2-1;
+			const auto u2nu=1-2*nu;
+			const auto q=q_(e);
+			const auto S11 = 3./(8*(1-nu))*e2/e21+q/4/(1-nu)*(u2nu-9./4/e21);
+			const auto S12= 1./4/(1-nu)*(e2/2/e21-q*(u2nu+3./4/e21));
+			const auto S13= 1./2/(1-nu)*(-e2/e21+q/2*(3.*e2/e21-u2nu));
+			const auto S31= 1./2/(1-nu)*(-u2nu-1./e21+q*(u2nu+3./2/e21));
+			const auto S33= 1./2/(1-nu)*(u2nu+(3.*e2-1.)/e21-q*(u2nu+3.*e2/e21));
+			const auto S44= 2./4/(1-nu)*(e2/2/e21+q*(u2nu-3./4/e21));
+			const auto S55= 2./4/(1-nu)*(u2nu-(e2+1.)/e21-q/2*(u2nu-3.*(e2+1)/e21));
```

Comment:
Use integers for constant values

Reply:
it seems that I must write for example (1-nu)/2 because 1/2/(1-nu) returns 0